theseus/labs/lie/lie_tensor.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from enum import Enum
+from typing import List, Tuple, Union
+
+import torch
+
+from theseus.labs.lie.functional.lie_group import UnaryOperatorOpFnType
+from theseus.labs.lie.functional import se3 as _se3_base
+from theseus.labs.lie.functional import so3 as _so3_base
+
+
+TensorType = Union[torch.Tensor, "LieTensor"]
+
+
+class LieType(Enum):
+    SE3 = 0
+    SO3 = 1
+
+
+SE3 = LieType.SE3
+SO3 = LieType.SO3
+
+_JFnReturnType = Tuple[List[torch.Tensor], TensorType]
+
+
+class LieTensor:
+    def __init__(self, data_tensor: torch.Tensor, ltype: LieType):
+        self._t = data_tensor
+        self._fn_lib = {
+            SE3: _se3_base,
+            SO3: _so3_base,
+        }[ltype]
```

Comment:
This could live outside the `LieTensor.__init__` function too. Not sure if there is advantage to doing it though.

Reply:
I think that defeats the purpose of having a separate `lie.functional` package for users to get lie group ops w/o using `LieTensor`. We could consider removing the "functional" subpackage, and moving that stuff one level up, but I like having the clear separation. 

Also, I have this idea that `lie.se3` and `lie.so3`, etc. would actually refer to Lie Algebra types, and `lie.SE3` and `lie.SO3` the corresponding group types (I think @maurimo was the first one that floated this). Although this is more of a loose thought right now, not sure if we'll need this.

Reply:
Oh good point. I missed the `[ltype]` and was thinking this dict is a constant.

Yeah I agree we should reserve both the `lie.SE3` and `lie.se3` versions in case we do decide to add the algebra support later. We need to switch the current namespace to uppercase since lower case is algebra, right?

Reply:
For consistency we could rename `_se3_base` -> `_se3_impl`?

Reply:
Changed.